### PR TITLE
pip-tools

### PR DESCRIPTION
**Pull request by silby**
pip-tools [1] is a tool for managing dependencies for Python
_applications_. (That is, it's less suitable for shared libraries
intended primarily for installation with pip.) It allows the developer
to specify _abstract_ Python dependencies (where the required version
isn't fixed, or is set to a range of possible versions), and still have
successive versions deployed with a tracked and locked set of _concrete_
dependencies. Unlike Pipenv, it doesn't take over management of virtual
environments; the developer must manage venvs for development and
deployment with other tools.

This commit modifies our example dockerfile to use pip-tools to install
our pinned dependencies from a requirements.txt file maintained at
development time. This is based on requirements.in, which uses the
standard requirements.txt syntax to specify dependencies, without
necessarily specifying an exact version.

Here I'm using the PEP 440 "compatible release" specifier to inform
pip-compile that it can depend on any version that is at least the
version number given (current major.minor versions at time of this
writing) other than a major version bump. Quick inspection of the recent
release history of the projects we depend on directly suggests that they
all adhere to the semver-esque convention of only introducing breaking
changes with major version upgrades, so for these dependencies, the
constraint is sensible. This may not be true for all packages we
directly depend on!

This means that to upgrade to a new major version of a package, we will
have to change requirements.in to bump it, which means we'll have to
somehow be aware of when there's a new major version. We might want to
look for or build a tool that will notify us.

This also introduces some changes to the Dockerfile. Instead of copying
a venv from the test stage container to the deployed container, we we
just run pip-sync install again when building the deployment container.
A few layers of the base image are built up and used as the starting
point for both the test and deployment stages. (Making a runas user and
home directory.) The application itself is no longer built or installed
at all using setuptools or pip.

Since pip-compile handles the resolution of abstract to concrete
dependencies at _development_ time, we no longer need to update
requirements.txt with the frozen deps ourselves.

The tox configuration, like the Dockerfile build, now doesn't build or
install the application as a package (skip_install), so the two
environments should match more closely.

The setup.py and setup.cfg files are no longer required, since no part
of the build system builds our app as a python dist.

The volume mounted by docker-compose is corrected to work as intended.
(I'm not certain it was working before.)

[1]: https://github.com/jazzband/pip-tools

### CODE REVIEW

**Comment by silby** [comment]
At this point, I'm happier with pip-tools than Pipenv. It does less, uses requirements.txt directly so it interacts in more obvious ways with other tools, and may be perceptually faster.

**Comment by silby** [comment]
needs a readme update

**Comment by benroy73** [comment]
pip-tools seems like a good thing for us right now and looks good to me.

One question for you and the team, is do we want to drop the setup.py in general or just in this example because it is a bespoke flask app? I believe that dropping setup.py makes our app not installable with pip or able to be distributed via pypi.

**Comment by silby** [comment]
I don't think there's any need for our flask apps to be installable or distributable, especially given the way we're deploying them. If we ever needed an app to be installable, we could add `setup.py` back in, but there's no reason to maintain it for apps otherwise with the present build and deployment strategy.

On the contrary, shared libraries _should_ be installable (such as `tools-auth` and `tools-constants` now are). We may eventually have a deployment strategy for command line tools to managed servers where we `pip install` them from a Python package, and `setuptools` is used by push deployments currently, so we can leave it in place there.

**Comment by benroy73** [comment]
Thanks, I think that's a good distinction.

File: Dockerfile [hunk 1]
```diff
@@ -10,47 +10,34 @@ USER toolop
 WORKDIR /home/toolop
 COPY . app
 
+WORKDIR /home/toolop
```
Comment (benroy73) [inline Dockerfile:13]: I think this line isn't needed.

Comment (silby) [reply Dockerfile:13]: sure looks that way.

**Pull request merged**